cmd/cstor-pool-mgmt/controller/pool-controller/handler.go

```diff
@@ -525,3 +545,77 @@ func (c *CStorPoolController) getDeviceIDs(csp *apis.CStorPool) ([]string, error
 	// TODO: Add error handling
 	return pool.GetDeviceIDs(csp)
 }
+
+func (c *CStorPoolController) upgrade(csp *apis.CStorPool) (*apis.CStorPool, error) {
+	var err error
+	if csp.VersionDetails.Current != csp.VersionDetails.Desired {
+		if !isCurrentVersionValid(csp) {
+			return nil, errors.Errorf("invalid current version %s", csp.VersionDetails.Current)
+		}
+		if !isDesiredVersionValid(csp) {
+			return nil, errors.Errorf("invalid desired version %s", csp.VersionDetails.Desired)
+		}
+		path := strings.Split(csp.VersionDetails.Current, "-")[0] + "-" +
+			strings.Split(csp.VersionDetails.Desired, "-")[0]
+		u := &upgradeParams{
+			csp:    csp,
+			client: c.clientset,
+		}
+		csp, err = upgradeMap[path](u)
+		if err != nil {
+			return csp, err
+		}
+		csp.VersionDetails.Current = csp.VersionDetails.Desired
+		csp, err = c.clientset.OpenebsV1alpha1().CStorPools().Update(csp)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to update SPC with versionDetails")
+		}
+		return csp, nil
+	}
+	return csp, nil
+}
+
+// populateVersion assigns VersionDetails for old csp object
+func (c *CStorPoolController) populateVersion(csp *apis.CStorPool) (
+	*apis.CStorPool, error,
+) {
+	v := csp.Labels[string(apis.OpenEBSVersionKey)]
+	// 1.3.0 onwards new CSP will have the field populated during creation
+	if v < "1.3.0" && csp.VersionDetails.Current == "" {
+		csp.VersionDetails.Current = v
+		csp.VersionDetails.Desired = v
+		obj, err := c.clientset.OpenebsV1alpha1().CStorPools().
+			Update(csp)
```

Comment:
We can look for patch as well , but can be done in upcoming PR's.

Reply:
In long run update would be better and we want to make the upgrade functions consistent for all resources.